sedaprotocol/seda-overlay
# Custom WASM downloading & execution

Implementation of custom WASM downloading & execution.

## Motivation

This allows users to create Data Requests with custom WASM binaries.

## Explanation of Changes

* Implement gRPC transparent encoder for sending and receiving arbitrary data
* Implement protobuf
* Before executing a DR the node checks if the `dr_binary_id` exists. If not it will try to download the binary from the SEDA chain and then cache it locally. 
* If the binary exists it will execute the binary and wait for its result. Once the result is available the node will post the results on chain

## Testing



This assumes there is already a contract deployed and configured for the node. The node needs to be registered with the oracle contract.

1. Start the node
2. If the chain got reset you need to build and upload the simple-dr. See section to build and upload simple-dr
3. Go to https://codesandbox.io/s/web3-react-v6-forked-5wc3yq?file=/src/index.js to generate a new data request (make sure to change something in the DR to not generate duplicates, best is to add a byte to `memo`, which has no effect on the execution)
4. Look in the console and find the `post_data_request` JSON.  (`{"post_data_request":{"posted_dr":{"dr_id": ... }}}`)
5. Copy the JSON and go to Ping Pub Explorer
6. Go to Cosmwasm and find the oracle contract that you are currently using
7. Click execute and paste the JSON as args
8. Sign and execute

The node will now pickup the DR and execute it. It will take care of the downloading and executing


### Simple-dr building & uploading
NOTICE: If the chain didn't reset i've already upload the latest simple dr binary: `006fd965caffcb4ec8fac83a192ec8ddbe3da20083621d7f8866bed4086fdee4` you don't have to do this then.

1. Go to the folder `wasm/test/simple-dr`
2. Execute the command `cargo build --release --target wasm32-wasi`
10. Take the `simple-dr.wasm` from target/wasm32-wasi/release and move it to a location where you have access to the latest `seda-chaind` (Can be inside the `seda-chain` project)
11. Execute the command `./seda-chaind tx wasm-storage store-data-request-wasm ./simple-dr.wasm --wasm-type data-request --from <YOUR_ADDRESS> --node tcp://18.169.59.167:26657 --gas 9647711` Where ofcourse the `./simple-dr.wasm` is accessible aswell as `./seda-chaind`
12. On pingpub look for your transaction and in the logs find Something like: `{"key":"hash","value":"\"006fd965caffcb4ec8fac83a192ec8ddbe3da20083621d7f8866bed4086fdee4\""}`

## Related PRs and Issues



closes #37 #12
